types/src/dataio.rs

```diff
@@ -1,36 +1,56 @@
+/// AlephBFT internally calls [`DataProvider::get_data`] whenever a new unit is created and data
+/// needs to be placed inside.
 ///
-/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/aleph_bft_api.html for a discussion
-/// and examples of how this trait can be implemented.
+/// We refer to the documentation
+/// https://cardinal-cryptography.github.io/AlephBFT/aleph_bft_api.html for a discussion and
+/// examples of how this trait can be implemented.
 #[async_trait]
-pub trait DataProvider<Data>: Sync + Send + 'static {
-    /// Outputs a new data item to be ordered
-    async fn get_data(&mut self) -> Option<Data>;
+pub trait DataProvider: Sync + Send + 'static {
+    /// Type of data returned by this provider.
+    type Output: Data;
+    /// Outputs a new data item to be ordered.
+    async fn get_data(&mut self) -> Option<Self::Output>;
 }
 
 /// The source of finalization of the units that consensus produces.
 ///
-/// The [`FinalizationHandler::data_finalized`] method is called whenever a piece of data input to the algorithm
-/// using [`DataProvider::get_data`] has been finalized, in order of finalization.
-pub trait FinalizationHandler<Data>: Sync + Send + 'static {
+/// The [`FinalizationHandler::data_finalized`] method is called whenever a piece of data input
+/// to the algorithm using [`DataProvider::get_data`] has been finalized, in order of finalization.
+pub trait FinalizationHandler<D>: Sync + Send + 'static {
     /// Data, provided by [DataProvider::get_data], has been finalized.
     /// The calls to this function follow the order of finalization.
-    fn data_finalized(&mut self, data: Data);
-    /// A unit has been finalized. You can overwrite the default implementation for advanced finalization handling
-    /// in which case the method [`FinalizationHandler::data_finalized`] will not be called anymore if a unit is finalized.
-    /// Please note that this interface is less stable as it exposes intrinsics which migh be subject to change.
-    /// Do not implement this method and only implement [`FinalizationHandler::data_finalized`] unless you
-    /// absolutely know what you are doing.
-    fn unit_finalized(&mut self, _creator: NodeIndex, _round: Round, data: Option<Data>) {
-        if let Some(d) = data {
-            self.data_finalized(d);
+    fn data_finalized(&mut self, data: D);
+}
+
+/// Represents state of the main internal data structure of AlephBFT (i.e. direct acyclic graph) used for
+/// achieving consensus.
+///
+/// Instances of this type are returned indirectly by [`member::run_session_for_units`] method using the
+/// [`UnitFinalizationHandler`]. This way it allows to reconstruct the DAG's structure used by AlephBFT,
+/// which can be then used for example for the purpose of node's performance evaluation.
+pub struct OrderedUnit<D: Data, H: Hasher> {
+    pub data: Option<D>,
+    pub parents: NodeMap<H::Hash>,
```

Comment:
Probably better to return a `Vec` of hashes here instead – this will always represent parents and the API will be more stable.

Reply:
changed